### ISSUE TITLE
marketo_ma_data can become populated by fields with blank keys

### ISSUE THREAD

**Issue by mccrodp** [MarketoMA/marketo_ma#15]
This bug can can cause "Field not found" errors on syncing with Marketo.

See `= 'Developer'` below.

<img width="528" alt="screen shot 2016-11-04 at 08 38 51" src="https://cloud.githubusercontent.com/assets/559938/20000731/e2fafeb4-a270-11e6-9b77-32f7391a6aea.png">

**Comment by jyokum** [comment]
Do you have steps to recreate this? In 7.x-2.x fields are no longer managed in a text field which was always a potential problem. Curious how a blank key could make its way into the data with the way fields are managed now.

**Comment by mccrodp** [comment]
I probably don't have steps that helpful as I'm working on a project that previously used Marketo REST. It has a large enough DB, so I'm not 100% sure what will be relevant. However, I don't think things overlapped in the modules re: variable names or in the DB (separate tables, etc).

So, only info I have is:
- A multi page webform with many fields setup
- Field Definitions fetched from Marketo
- Fields did not default to none as they would now after merging: https://github.com/MarketoMA/marketo_ma/pull/14 and were saved with mappings to `Marketo Data.com Status()` field by default.
- Mapped textfield field with issue has label "Title" and machine name `agency_author_title`

I think it's safest to still check to ensure that non blank keys are checked, as perhaps in more custom Webform, things can get a bit messed up. I'm having no issues after https://github.com/MarketoMA/marketo_ma/pull/16 even though I left potentially incorrect field mappings in place.

It's strange though as many fields were mapped to this `Marketo Data.com Status()` field, but only the last component field appears in the screenhot above when debugging. The other components/fields don't seem to be mapped  and don't appear here.

**Comment by mccrodp** [comment]
Hey @jyokum. Have some more data, didn't find the exact root of the problem, but I believe I have enough info to show the point.

Sorry for screens, but easiest way to illustrate.

The fieldmap can have blank keys:
<img width="625" alt="blank-field-key" src="https://cloud.githubusercontent.com/assets/559938/20051122/32116104-a4c5-11e6-94e8-ffa52f79b877.png">

I believe due to before the merge of https://github.com/MarketoMA/marketo_ma/pull/14 that if the form was saved, some of these blank keys can be mapped (I tested without this PR included). Then it steps past the `$key != MARKETO_MA_WEBFORM_COMPONENT_NONE` check and we can have null keys:

<img width="623" alt="blank-field-key-2" src="https://cloud.githubusercontent.com/assets/559938/20051188/a52e99fe-a4c5-11e6-99b0-cea6886f772b.png">

In my case the last one of these incorrectly mapped keys (cid 29) gets assigned it's field value "Developer".:

<img width="624" alt="blank-field-key-3" src="https://cloud.githubusercontent.com/assets/559938/20051248/f9d026d0-a4c5-11e6-8f00-a53010ec593a.png">

This causes an issue in `syncLead`, only the fields where only `email` field gets synced, the other fields get ignored and we get a field not found error.

I am going to test now if we avoid this issue with the PR https://github.com/MarketoMA/marketo_ma/pull/14, but I imagine not for the specific form I'm testing with, as this webform settings are perhaps mapping to this `Marketo Data.com Status()` field already and this causes issues somehow.

Do you agree that we should disallow and check for blank keys in the `$fieldmap` or do you want to dig deeper into this first before merging the PR?

**Comment by mccrodp** [comment]
Perhaps `_marketo_ma_webform_get_mapped_fields` should also ensure that it only returns non nulled values for the keys.

``` php
$result = db_select(MARKETO_MA_SCHEMA_WEBFORM_COMPONENT)
    ->fields(MARKETO_MA_SCHEMA_WEBFORM_COMPONENT)
    ->condition('nid', $nid)
    ->condition(MARKETO_MA_WEBFORM_COMPONENT_KEY, NULL,'!=') // Add this line.
    ->execute()
    ->fetchAll();
```

I'll add the above to this PR https://github.com/MarketoMA/marketo_ma/pull/14

**Comment by mccrodp** [comment]
I changed the method to ensure that the `marketo_ma_key` cannot have null value, we do not call db_merge without removing null keys first and we do not try to `syncLead` with any null keyed fields: https://github.com/MarketoMA/marketo_ma/pull/16/files

**Comment by jyokum** [comment]
Do you have fields in the marketo_ma_lead_fields table which have blank keys? I'd really like to figure out what caused this to happen.

**Comment by mccrodp** [comment]
Yep, you nailed it.

> 123   Marketo Data.com Status string  50  NULL    0   Marketo Jigsaw Contact Status   0   0

`restName` is `NULL` for `Marketo Data.com Status`.

So, these NULLs are possible if we have a NULL `restName` at least. Let's agree on a way to prevent it for `restName` and potentially `soapName` also where munchkin API is concerned. Check out what I have and let me know if you think there's a better way.

**Comment by jyokum** [comment]
Take a look at what fields are coming back from Marketo when marketo_ma.rest.inc getFields() is executed. Is that key blank in the Marketo response or possibly getting lost in prepareFields($response)?

Module config > Field Definition > Retrieve from Marketo will trigger it.

**Comment by jyokum** [comment]
I think I see what's happening now and we'll need to do more than just check for a blank rest key. We should probably change the webform component mapping behavior to map to the field ID rather than the restName (or soapName) directly. From there, retrieve the appropriate field name depending on the tracking method when the form is submitted.

**Comment by mccrodp** [comment]
It's a SOAP only field, there are some REST only ones also, so `NULL` values in either API are to be expected. I believe `prepareFields` is functioning as expected, the `rest` entry is just not included in the response for this field.

> {"id":33,"displayName":"Marketo Data.com Status","dataType":"string","length":50,"soap":{"name":"Marketo Jigsaw Company Status","readOnly":false}}

**Comment by mccrodp** [comment]
> We should probably change the webform component mapping behavior to map to the field ID rather than the restName (or soapName) directly.

Ahh I didn't notice it was mapped on the rest/soap key rather than the ID. Thought you took what was in the `marketo_rest` module over, but perhaps not for this part. Yes, we need to do that.

Ok, so if we do that it allows to swap between REST / SOAP on a form seamlessly, but only for the fields that have a value for that tracking method. 

We should exclude fields that do not have a fieldname for the current tracking method on the "Marketo" webform settings form tab. But, if tracking method is changed, it is still possible to have a NULL fieldname.

What would you prefer to do when a field is mapped to a `NULL` fieldname for a particular tracking method?

I had not accounted for the above in `marketo_rest` either, but I will take a look to see if anything we can reuse from there for the key mapping. I believe I used `marketo_ma_key` equivalent as the field id from Marketo, shall we do the same here or would you prefer to do a different way?

**Comment by mccrodp** [comment]
Actually, I'm wrong, I used a _field_id_ const and mapped fields on that, I didn't reuse the ~ `ma_key` const 
- https://github.com/mccrodp/marketo_rest/blob/7.x-1.x/webform/marketo_rest_webform.install#L58
- https://github.com/mccrodp/marketo_rest/blob/7.x-1.x/webform/marketo_rest_webform.module#L151

**Comment by jyokum** [comment]
Merged #22 

Issue should be resolved now that we're using IDs to reference fields rather that the REST or SOAP names which could potentially be blank.

**Comment by mccrodp** [comment]
Yep, should be sorted now, thanks for your work on getting this sorted via mapping PR.
